### PR TITLE
Game of Life tutorial

### PR DESCRIPTION
**Pull request by fitzgen**
I still need to write the code size profiling chapter and the publishing with wasm-pack chapter, but I think we can pipeline that with review of the first few chapters :)

All the complete code for each chapter is here: https://github.com/fitzgen/wasm_game_of_life

Would love some feedback from anyone who wants to give it!

I hope that this will eventually become the main book, and everything else will be appendices/references at the end of this extended tutorial, but this can happen later separately, and we can discuss this at that time.

### CODE REVIEW

File: src/game-of-life/setup.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+### `wasm-bindgen`
+
+`wasm-bindgen` generates bidirectional bindings to and from JavaScript for Rust
+and WebAssembly.
+
+Install `wasm-bindgen` with this command:
+
+```
+rustup run nightly cargo install wasm-bindgen --version 0.2.0
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: I'd recommend `cargo +nightly` here instead of `rustup run`, and the `--version` may also wish to be omitted. Unfortunately as new version of wasm-bindgen-the-crate are published older versions of `wasm-bindgen` won't work (various issues there), so it's probably best if we generally make sure everything stays up-to-date.

File: src/game-of-life/setup.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+├── Cargo.lock
+├── Cargo.toml
+├── index.html
+├── index.js
+├── package.json
+├── package-lock.json
+├── src
+│   └── lib.rs
+├── wasm_game_of_life_bg.wasm
+├── wasm_game_of_life.js
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: I think we might benefit from some gitignores in the branches of this repo perhaps? I think the generate `wasm_game_of_life*` files may want to be ignored by default as well as `Cargo.lock`.

Although actually if we keep `Cargo.lock` that may solve the `--vers` thing above as well!

File: src/game-of-life/setup.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+The project template contains a "hello world" program. It comes pre-configured
+with sane defaults, so you can quickly build, integrate, and package your code
+for the Web.
+
+Clone this tutorials code repository, enter its directory, and checkout the
+`chapter-zero` branch:
+
+```text
+git clone https://github.com/fitzgen/wasm_game_of_life.git
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: Want to move this into the `rustwasm` org as well?

File: src/game-of-life/setup.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+</html>
+```
+
+### `index.js`
+
+The `index.js` is the main entry point for our Web page's JavaScript. It imports
+the project's WebAssemly module, and calls the module's `greet` function.
+
+```js
+import * as wasm from "./wasm_game_of_life";
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: I'm not actually sure what the best practice is here but at least in the Glimmer codebase using a glob import is pretty rare and this'd instead be:

```js
import { greet } from "./wasm_game_of_life";
```

not sure which is best and/or more idiomatic?

Comment (alexcrichton) [reply src/game-of-life/setup.md]: Some other thoughts! The name `wasm` here isn't necessarily 100% accurate too in the sense that it's actually JS! I wonder if that also means that `game_of_life` may be a more appropriate module name in the sense that it's an implementation ideally that wasm is used under the hood

File: src/game-of-life/setup.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+wasm.greet("Rust and WebAssembly");
+```
+
+### `src/lib.rs`
+
+The `src/lib.rs` file is the root of the Rust crate that we are compiling to
+WebAssembly. It uses `wasm_bindgen` to interface with JavaScript. It imports the
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: Hyperlink wasm_bindgen?

File: src/game-of-life/setup.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+First, ensure that the JavaScript build dependencies are installed for this
+project:
+
+```text
+npm install
+```
+
+This command only needs to be run once, and will install the `webpack`
+JavaScript bundler and its development server.
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: It might be good to have as a note here that Webpack isn't required, it's just the bundler chosen for ease of use here. (I think others have been scared off in the past when it looks like we might require webpack)

File: src/game-of-life/setup.md (hunk 2 above)
Comment (alexcrichton) [reply src/game-of-life/setup.md]: Ok actually due to the fact that this still requires nightly I'd recommend dropping `Cargo.lock`, it looks like it's pinned to an older version of `proc-macro2` which no longer compiles.

File: src/game-of-life/setup.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+To build the Rust crate as WebAssembly and generate the `wasm_bindgen` glue, run
+this command:
+
+```text
+npm run build
+```
+
+The first build may take a little while, but subsequent builds will be much
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: Right now this builds in release mode by default, but I wonder if for a better intro experience there could be to targets like `npm run build` and `npm run build-release`?

File: src/game-of-life/setup.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+Navigate your Web browser to [http://localhost:8080/](http://localhost:8080/)
+and you should be greeted with an alert message:
+
+[![Screenshot of the "Hello, Rust and WebAssembly!" Web page alert](/images/game-of-life/setup.png)](/images/game-of-life/setup.png)
+
+Anytime you make changes and want them reflected on
+[http://localhost:8080/](http://localhost:8080/), just re-run the `npm run
+build` command.
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: I changed the Rust code and reran `npm run build`, but on refreshing the browser it didn't change anything :(

Debugging a bit it looks like `npm run serve` (aka `webpack-dev-server`) needs to be restarted right now whenever the wasm file changes. It does work correctly, however, if the JS changes!

File: src/game-of-life/rules.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+Consider the following initial universe:
+
+<style>
+.universe {
+    font-family: monospace;
+}
+</style>
+
+<table class="universe">
+<tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td></tr>
```
Comment (alexcrichton) [inline src/game-of-life/rules.md]: The table columns/rows were a little hard for me to see on Firefox on Windows:

capture

is there a way to attach styling here to help that?

File: src/game-of-life/implementing.md [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+dead cell and `1` is a live cell.
+
+Here is what a 4 by 4 universe looks like in memory:
+
+<!-- For whatever reason, the more specific monospace fonts break with the
+following box-drawing characters, hence the manual style.-->
+<pre style="font-family: monospace">
+Indices:   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15
+         ┏━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┯━━━┓
+  Array: ┃   │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   ┃
```
Comment (alexcrichton) [inline src/game-of-life/implementing.md]: Ah unfortunately this didn't show up that well for me in a browser:


capture

File: src/game-of-life/implementing.md [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+addition.
+
+Next, let's define the universe. The universe has a width and a height, and a
+vector of cells of length `width * height`.
+
+```rust
+#[wasm_bindgen]
+pub struct Universe {
+    width: u32,
+    height: u32,
```
Comment (alexcrichton) [inline src/game-of-life/implementing.md]: I was playing around with various optimizations last night and one of the biggest optimizations is actually making these constant rather than runtime values. I'm curiuos, though, why they're specified here? I'd imagine these to be exported through `Universe::new` but it looks like they're constant there as well?

I'm totally fine leaving this as a possible optimization to implement later, was just curious!

File: src/game-of-life/implementing.md [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+const DEAD = 0;
+const ALIVE = 1;
+```
+
+The JavaScript runs in a `requestAnimationFrame` loop. On each iteration, it
+draws the current universe to the `<canvas>`, and then calls `Universe::tick`.
+
+```js
+// Construct the universe, and get its width and height.
+const universe = wasm.Universe.new();
```
Comment (alexcrichton) [inline src/game-of-life/implementing.md]: Similar to elsewhere this may be better done as just `Universe.new()` instead of glob-importing `wasm`

File: src/game-of-life/debugging.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+If you don't have debug symbols enabled, then the `"name"` section won't be
+present in the compiled `.wasm` binary, and stack traces will have function
+names like `wasm[42]` rather than
+`wasm_game_of_life::Universe::live_neighbor_count`.
+
+To enable debug symbols, ensure that you `debug = true` in the
+`[profile.release]` section of your `Cargo.toml`:
+
+```toml
+[profile.release]
```
Comment (alexcrichton) [inline src/game-of-life/debugging.md]: I was thinking elsewhere that we'd want to default to development mode rather than release mode which may want to tweak the wording of this section as well.

File: src/game-of-life/setup.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+new terminal lets us leave it running in the background, and doesn't block us
+from running other commands in the meantime. In the new terminal, run this
+command:
+
+```
+npm run serve
+```
+
+Navigate your Web browser to [http://localhost:8080/](http://localhost:8080/)
+and you should be greeted with an alert message:
```
Comment (alexcrichton) [inline src/game-of-life/setup.md]: Around here we may want a [caveat for Chrome users](https://github.com/rustwasm/wasm-bindgen/tree/master/examples/hello_world#caveat-for-chrome-users)

File: src/game-of-life/debugging.md [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+hypotheses about why our programs are buggy. On the Web, the `console.log`
+function is the way to log messages to the browser's developer tools console. We
+can use `wasm_bindgen` to import a reference to it, like this:
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(js_namespace = console)]
+    fn log(msg: &str);
+}
```
Comment (alexcrichton) [inline src/game-of-life/debugging.md]: I might recommend adding a macro as well like:

```rust
macro_rules! log {
    ($($t:tt)*) => (log(&format!($($t)*)))
}
```

(to be used like `println!`)

File: src/game-of-life/debugging.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+
+## Avoid the Need to Debug WebAssembly in the First Place
+
+While some bugs are specific to interfaceing JavaScript and WebAssembly,
+experience says that most are not. Try to reproduce bugs as normal Rust
+`#[test]` functions, where you can leverage your OS's mature native tooling when
+debugging. Use testing crates like [`quickcheck`][quickcheck] to exercise the
+interface you expose to JavaScript. Ultimately, you will have an easier time
+finding and fixing bugs if you can isolate them in a smaller test cases that
+don't require interacting with JavaScript.
```
Comment (alexcrichton) [inline src/game-of-life/debugging.md]: One thing this section doesn't go into much detail about is how you'd actually run the tests. This is related to the `cdylib` aspect today, but I wonder if we may want to provide a small bit of guidance about how to run tests here?

File: src/game-of-life/debugging.md [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+## Exercises
+
+* Add logging to the `tick` function that records the row and column of each
+  cell that transitioned states from live to dead or vice versa.
+
+* Introduce a `panic!()` in the `Universe::new` method. Inspect the panic's
+  backtrace in your Web browser's JavaScript debugger. Disable debug symbols,
+  rebuild, and inspect the stack trace again. Not as useful, is it?
+
+* Checkout the `chapter-one-with-bug` branch. Rebuild and reload the Web
```
Comment (alexcrichton) [inline src/game-of-life/debugging.md]: It looks like this branch isn't preloaded with a `log` binding, but maybe that could be included for convenience?

File: src/game-of-life/debugging.md [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+  page. It should now be obvious that this branch's implementation contains a
+  bug, and every cell is apparently alive. This is a Real World(tm) bug that
+  your author made when initially creating the example code. Find the bug and
```
Comment (alexcrichton) [inline src/game-of-life/debugging.md]: This took me an embarrasingly long time to find

File: src/game-of-life/interactivity.md [hunk 19]
```diff
@@ -0,0 +1,173 @@
+# Adding Interactivity
+
+We will continue to explore the JavaScript and WebAssembly interface by adding
+some interactive features to our Game of Life implementation. We will enable
+users to toggle toggle whether a cell is alive or dead by clicking on it, and
+allow pausing the game, which makes drawing cell patterns a lot easier.
+
```
Comment (alexcrichton) [inline src/game-of-life/interactivity.md]: Could the `chapter-two` branch be mentioned here too?

File: src/game-of-life/interactivity.md [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+  const relativeLeft = event.pageX - event.target.offsetLeft;
+  const relativeTop = event.pageY - event.target.offsetTop;
+
+  const row = Math.min(Math.floor(relativeTop / (CELL_SIZE + 1)), height - 1);
+  const col = Math.min(Math.floor(relativeLeft / (CELL_SIZE + 1)), width - 1);
+
+  universe.toggle_cell(row, col);
+
+  drawCells();
+  drawGrid();
```
Comment (alexcrichton) [inline src/game-of-life/interactivity.md]: Something feels a bit funny with the implementation here I think. If I pause and then click it moves the universe forward one step I think?

Comment (alexcrichton) [reply src/game-of-life/interactivity.md]: (subsequent clicks don't move the universe, they just toggle)

File: src/game-of-life/interactivity.md [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+the click event's page-relative coordinates into a row and column, invoke the
+`toggle_cell` method, and finally redraw the scene.
+
+```js
+canvas.addEventListener("click", event => {
+  const relativeLeft = event.pageX - event.target.offsetLeft;
+  const relativeTop = event.pageY - event.target.offsetTop;
+
+  const row = Math.min(Math.floor(relativeTop / (CELL_SIZE + 1)), height - 1);
+  const col = Math.min(Math.floor(relativeLeft / (CELL_SIZE + 1)), width - 1);
```
Comment (alexcrichton) [inline src/game-of-life/interactivity.md]: I think this calculation may not be quite right (although I'm not sure why). When I change `CELL_SIZE` to 20 or something larger (5px is tiny on my screen) then clicking on a square changes a random square elsewhere in the grid.

File: src/game-of-life/time-profiling.md [hunk 22]
```diff
@@ -0,0 +1,573 @@
...
+display which functions are taking the most time with the usual kinds of
+visualizations like call trees and flame graphs. If you [build with debug
+symbols][symbols], then these profilers should display the Rust function names
+instead of something like `wasm-function[123]`. Note that these profilers
+*won't* show inlined functions, and since Rust and LLVM rely on inlining so
+heavily, the results might end up a bit perplexing.
+
+[symbols]: /game-of-life/debugging.html#building-with-debug-symbols
+
+[![Screenshot of profiler with Rust symbols](/images/game-of-life/profiler-with-rust-names.png)](/images/game-of-life/profiler-with-rust-names.png)
```
Comment (alexcrichton) [inline src/game-of-life/time-profiling.md]: Interestingly this has a bunch of mangled symbols and for whatever reason demangling is [explicitly disabled](https://github.com/rustwasm/wasm-bindgen/blob/d1a4bffb3a767a69625130051e2f2f384b42c9b5/crates/cli-support/src/js.rs#L1170-L1178). What do you think about switching `wasm-bindgen` to demangle by default?

File: src/game-of-life/debugging.md (hunk 18 above)
Comment (fitzgen) [reply src/game-of-life/debugging.md]: me too :-p

**Comment by mgattozzi** [comment]
Just commenting that I did see this PR and I'll go through the whole thing when I have the time and then I'll leave feedback!

File: src/game-of-life/setup.md (hunk 3 above)
Comment (fitzgen) [reply src/game-of-life/setup.md]: Yes, will do when we merge this PR.

File: src/game-of-life/setup.md (hunk 8 above)
Comment (fitzgen) [reply src/game-of-life/setup.md]: I'll file an upstream bug. I've run into this as well.

File: src/game-of-life/implementing.md (hunk 10 above)
Comment (fitzgen) [reply src/game-of-life/implementing.md]: Dang... will have to figure something out. Suggestions?

File: src/game-of-life/rules.md (hunk 9 above)
Comment (fitzgen) [reply src/game-of-life/rules.md]: I guess we can add custom CSS to make the borders darker?

File: src/game-of-life/implementing.md (hunk 11 above)
Comment (fitzgen) [reply src/game-of-life/implementing.md]: Fascinating!

I figured that at some point in time it could be runtime configurable, but never ended up implementing that or making it an exercise.

File: src/game-of-life/debugging.md (hunk 16 above)
Comment (fitzgen) [reply src/game-of-life/debugging.md]: I think we should just fix that issue where you can't make the same crate compile to wasm, have tests, and have benches.

File: src/game-of-life/interactivity.md (hunk 19 above)
Comment (fitzgen) [reply src/game-of-life/interactivity.md]: I added a mention of it right before the exercises, similar to `implementing.md`.

File: src/game-of-life/interactivity.md (hunk 20 above)
Comment (fitzgen) [reply src/game-of-life/interactivity.md]: I know what this is -- I'll fix it.

File: src/game-of-life/time-profiling.md (hunk 22 above)
Comment (fitzgen) [reply src/game-of-life/time-profiling.md]: Yeah, I think that would be helpful.

File: src/game-of-life/interactivity.md (hunk 21 above)
Comment (fitzgen) [reply src/game-of-life/interactivity.md]: Huh. I'll dig into it.

File: src/game-of-life/setup.md (hunk 1 above)
Comment (alexcrichton) [reply src/game-of-life/setup.md]: I originally thought that as well yeah but I think that the fact that it all requires nightly is a reason to not lock down in my mind. That means that whenever anything breaks we have to publish a fix *and* update everything here, and otherwise I think the likelihood of newer versions breaking this would be pretty low. Could go either way though!

File: src/game-of-life/setup.md (hunk 2 above)
Comment (alexcrichton) [reply src/game-of-life/setup.md]: Oh excellent point! Perhaps a dedicated `gh-pages` branch though which aggregates all chapters?

File: src/game-of-life/implementing.md (hunk 10 above)
Comment (alexcrichton) [reply src/game-of-life/implementing.md]: Maybe a code block could be used and that may have better luck?

File: src/game-of-life/implementing.md [hunk 23]
```diff
@@ -0,0 +1,493 @@
...
+We will implement the third option.
+
+### Interfacing Rust and JavaScript
+
+> ⚡ This is one of the most important concepts to understand and take away from
+> this tutorial!
+
+JavaScript's garbage-collected heap — where `Object`s, `Array`s, and DOM nodes
+are allocated — is distinct from WebAssembly's linear memory space, where our
+Rust values live. WebAssembly (currently) has no direct access to the
```
Comment (killercup) [inline src/game-of-life/implementing.md]: maybe make it "currently (April 2018)"?

File: src/game-of-life/implementing.md [hunk 24]
```diff
@@ -0,0 +1,493 @@
...
+garbage-collected heap. JavaScript, on the other hand, can read and write to the
+WebAssembly linear memory space, but only as an `ArrayBuffer` of scalar values
```
Comment (killercup) [inline src/game-of-life/implementing.md]: link to <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer> maybe?

File: src/game-of-life/implementing.md [hunk 25]
```diff
@@ -0,0 +1,493 @@
...
+(`u8`, `i32`, `f64`, etc...). WebAssembly functions also take and return scalar
+values. These are the building blocks upon which all WebAssembly and JavaScript
+communication is built.
```
Comment (killercup) [inline src/game-of-life/implementing.md]: > building blocks upon which […] are built

How about s/built/assembled?

File: src/game-of-life/implementing.md [hunk 26]
```diff
@@ -0,0 +1,493 @@
...
+
+2. **Minimizing serializing and deserializing.** Similar to copies, serializing
+   and deserializing also imposes overhead, and often imposes copying as
+   well. If we can pass opaque handles to a data structure — instead of
+   serializing it on one side, copying it into some known location in the
+   WebAssembly linear memory, and deserializing on the other side — we can often
+   reduce a lot of overhead. `wasm_bindgen` helps us define and work with opaque
+   handles to JavaScript `Object`s or boxed Rust structures.
+
+3. **Minimizing the number of times JavaScript calls an exported WebAssembly
```
Comment (killercup) [inline src/game-of-life/implementing.md]: compare to syscalls in native binaries?

File: src/game-of-life/implementing.md [hunk 27]
```diff
@@ -0,0 +1,493 @@
...
+
+To find the array index of the cell at a given row and column in the universe,
+we can use this formula:
+
+```text
+index(row, column, universe) = row * height(universe) + column
+```
+
+To expose the universe's cells to JavaScript, we return a pointer to the start
+of the array. The JavaScript knows the width and height of the universe, and can
```
Comment (killercup) [inline src/game-of-life/implementing.md]: "The JavaScript [code] knows"

File: src/game-of-life/implementing.md [hunk 28]
```diff
@@ -0,0 +1,493 @@
...
+*Another viable design alternative would be for Rust to return a list of every
+cell that changed states after each tick, instead of exposing the whole universe
+to JavaScript. This way, JavaScript wouldn't need to iterate over the whole
+universe when rendering, only the relevant subset. The trade off is that this
+delta-based design is slightly more difficult to implement.*
+
+## Rust Implementation
+
+Let's begin by removing the `alert` import and `greet` function from
+`src/lib.rs`, and replacing them with a type definition for cells:
```
Comment (killercup) [inline src/game-of-life/implementing.md]: maybe start with something like "In the Setup chapter, we cloned the project template…"

File: src/game-of-life/implementing.md [hunk 29]
```diff
@@ -0,0 +1,493 @@
...
+```rust
+#[wasm_bindgen]
+pub struct Universe {
+    width: u32,
+    height: u32,
+    cells: Vec<Cell>,
+}
+```
+
+To access the cell at a given row and column, we use translate the row and
```
Comment (killercup) [inline src/game-of-life/implementing.md]: ~~use~~

File: src/game-of-life/implementing.md [hunk 30]
```diff
@@ -0,0 +1,493 @@
...
+        self.cells.as_ptr()
+    }
+}
+```
+
+With that, the Rust half of our Game of Life implementation is complete!
+
+## Rendering with JavaScript
+
+First, let's add the `<canvas>` we will render into to `index.html`. It should
```
Comment (killercup) [inline src/game-of-life/implementing.md]: woah, starting with canvas here? I'd probably use text and have the same style as I proposed above (◻️ and ◼️)

(this is also how far I've read so far so sorry if it all makes sense later)

File: src/game-of-life/rules.md (hunk 9 above)
Comment (killercup) [reply src/game-of-life/rules.md]: I'd probably try to go for a text representation, using ◻️ and ◼️ for example.

File: src/game-of-life/implementing.md (hunk 10 above)
Comment (killercup) [reply src/game-of-life/implementing.md]: 2min in [asciiflow](http://asciiflow.com):

```plain
Indices: | 0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15 |
         +---+---+---+---+---+---+-------+---------------+---------------+
  Array: |   |   |   |   |   |   |   │   |   │   │   │   |   │   │   │   |
         +---+---+---+---+---+---+-------+---------------+---------------+
   Rows: |     row 0     |     row 1     |     row 2     |     row 3     |
```

If you want something prettier, I'd try and make a SVG.

File: src/game-of-life/implementing.md (hunk 30 above)
Comment (killercup) [reply src/game-of-life/implementing.md]: Just for fun, here's the rust debug impl:

```rust
use std::fmt;
impl fmt::Debug for Universe {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for line in self.cells.as_slice().chunks(self.width as usize) {
            for &cell in line {
                write!(f, "{}", if cell == Cell::Dead { "◻️" } else { "◼️" })?;
            }
            write!(f, "\n")?;
        }

        Ok(())
    }
}
```

File: src/game-of-life/setup.md (hunk 4 above)
Comment (fitzgen) [reply src/game-of-life/setup.md]: Done!

File: src/game-of-life/setup.md (hunk 5 above)
Comment (fitzgen) [reply src/game-of-life/setup.md]: I made the first reference to wasm-bindgen a link.

File: src/game-of-life/rules.md (hunk 9 above)
Comment (fitzgen) [reply src/game-of-life/rules.md]: Done!

File: src/game-of-life/debugging.md (hunk 13 above)
Comment (fitzgen) [reply src/game-of-life/debugging.md]: Done.

File: src/game-of-life/setup.md (hunk 14 above)
Comment (fitzgen) [reply src/game-of-life/setup.md]: Done.

File: src/game-of-life/debugging.md (hunk 15 above)
Comment (fitzgen) [reply src/game-of-life/debugging.md]: Done.

File: src/game-of-life/interactivity.md (hunk 20 above)
Comment (fitzgen) [reply src/game-of-life/interactivity.md]: Done.

File: src/game-of-life/interactivity.md (hunk 21 above)
Comment (fitzgen) [reply src/game-of-life/interactivity.md]: Fixed.

File: src/game-of-life/implementing.md (hunk 30 above)
Comment (fitzgen) [reply src/game-of-life/implementing.md]: Yes, I think it ended up working well to start with canvas. That's a nice, concise text rendering implementation, though :)

File: src/game-of-life/implementing.md (hunk 26 above)
Comment (fitzgen) [reply src/game-of-life/implementing.md]: Good call -- added a comment mentioning the similarities.

File: src/game-of-life/implementing.md (hunk 30 above)
Comment (killercup) [reply src/game-of-life/implementing.md]: Sorry, I was in a rush yesterday. Let me rephrase my original comment :)

It seems like the current guide takes a few steps at a time:

- It goes from "here's a Rust struct" (simple)
- via "let's read the actual memory" (woah what for JS programmers)
- to "draw stuff with the canvas API" (adds a lot of code; and uses and API that is quite simple but that may be unfamiliar to many JS programmers).

I can see this being a bit overwhelming to people who have never done any one of these three things. (I'm bad at judging this, though.)

If you were to start with the rust impl as a native binary to test that it works and then added wasm_bindgen, you could've at first returned `universe.to_string()` to JS, and in a later chapter switch to reading the actual memory instead of doing expensive string copying.

(Sadly making this work as a binary seems to involve changing the Cargo.toml to not compile a cdylib which is annoying.)

File: src/game-of-life/time-profiling.md [hunk 31]
```diff
@@ -0,0 +1,577 @@
...
+  where the `Universe` maintains two vectors, never frees either of them, and
+  never allocates new buffers in `tick`.
+
+* Implement the alternative, delta-based design from the "Implementing Life"
+  chapter, where the Rust code returns a list of cells that changed states to
+  JavaScript. Does this make rendering to `<canvas>` faster? Can you implement
+  this design without allocating a new list of deltas on every tick?
+
+* As our profiling has shown us, 2D `<canvas>` rendering is not particularly
+  fast. Replace the 2D canvas renderer with a WebGL renderer. How much faster is
```
Comment (alexcrichton) [inline src/game-of-life/time-profiling.md]: FWIW I was curious to learn about WebGL and try this out, and you have to *seriously* stress the system before something like this becomes necessary. I've gotten up to a 1024x1024 grid which can do up to thousands of ticks per frame and maintain 60fps.

File: src/game-of-life/code-size.md [hunk 32]
```diff
@@ -0,0 +1,280 @@
...
+find target/release -type f -name '*.ll'
+```
+
+#### References
+
+* [LLVM Language Reference Manual](https://llvm.org/docs/LangRef.html)
+
+## More Tools and Techniques
+
+Here is a collection of techniques you can apply to get smaller code sizes.
```
Comment (alexcrichton) [inline src/game-of-life/code-size.md]: It may be worth starting off this section commenting how most of the pieces above are tweaking the build or otherwise diagnosing where the size is coming from. This section is specifically about how you can change your code to diagnose the issues discovered during profiling.

File: src/game-of-life/code-size.md [hunk 33]
```diff
@@ -0,0 +1,280 @@
...
+
+### Avoid String Formatting
+
+`format!`, `to_string`, etc... can bring in a lot of code bloat. If possible,
+only do string formatting in debug mode, and in release mode use static strings.
+
+### Avoid Panicking
+
+This is definitely easier said than done, but tools like `svelte` and manually
+inspecting LLVM-IR can help you figure out which functions are panicking.
```
Comment (alexcrichton) [inline src/game-of-life/code-size.md]: It is probably worth commenting how things like `a[i]` or `b.unwrap()` are implicitly causing panics, and will contribute to code size bloat. A strategy I often use to work around this is to punt errors to wasm (aka boolean returns or something like that) rather than letting Rust panic

File: src/game-of-life/implementing.md (hunk 30 above)
Comment (fitzgen) [reply src/game-of-life/implementing.md]: I see what you're getting at.

I am less concerned about JS programmers being unfamiliar with canvas, than I am with Rust programmers being unfamiliar with it. But even there I am not super concerned (maybe I should be?)

I agree that the "actual memory" bits might be like a cold shower for JS hackers. I am more concerned here, but I also don't know how to ease into this more. More explanation and exposition before diving in?

I think we could skip the binary and start with sending a text representation across the boundary. This addresses the canvas unfamiliarity, but not the actual memory bits.

Ideas?

File: src/game-of-life/code-size.md [hunk 34]
```diff
@@ -0,0 +1,280 @@
...
+
+### Disable Debug Symbols
+
+The section containing function name strings for debugging takes up more space
+than you might think. If you aren't debugging or profiling, and want to generate
+a small `.wasm` binary, then disable these names with this `Cargo.toml` setting:
+
+```toml
+[profile.release]
+debug = false
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply src/game-of-life/code-size.md]: I have been unsure how to handle "defaults" since there are two sets of defaults:

1. Cargo defaults.
2. The template's opinionated defaults.

They don't always agree.

File: src/game-of-life/code-size.md [hunk 35]
```diff
@@ -0,0 +1,280 @@
...
+
+With the default release build configuration (without debug symbols), our
+WebAssembly binary is 240,605 bytes:
+
+```
+$ wc -c wasm_game_of_life_bg.wasm
+240605 wasm_game_of_life_bg.wasm
+```
+
+After enabling LTO, setting `opt-level = "z"`, and running `wasm-opt -Oz`, the
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply src/game-of-life/code-size.md]: Interesting! Will check...

File: src/game-of-life/time-profiling.md [hunk 36]
```diff
@@ -0,0 +1,577 @@
...
+
+* At this point, the next lowest hanging fruit for speeding up `Universe::tick`
+  is removing the allocation and free. Implement double buffering of cells,
+  where the `Universe` maintains two vectors, never frees either of them, and
+  never allocates new buffers in `tick`.
+
+* Implement the alternative, delta-based design from the "Implementing Life"
+  chapter, where the Rust code returns a list of cells that changed states to
+  JavaScript. Does this make rendering to `<canvas>` faster? Can you implement
+  this design without allocating a new list of deltas on every tick?
```
Comment (killercup) [inline src/game-of-life/time-profiling.md]: Another (albeit somewhat absurd) idea we talked about was bit-packing the cells: Right now we have 7 unused bits per cell! Packing 8 cell states into a byte, a 4096⨉4096 universe would instead of 16MB only need to allocate 2MB.

If we had to transfer this data to JS each time, or if we deal with huge universes, we could save a bunch of memory here. Assuming we have a universe width divisible by 8, we could also try and draw 8 cells at a time. This is obviously at the cost of algorithmic complexity, both at runtime and when writing the code. It might also open the door to some new fancy ways to quickly determine if entire blocks of cells are dead, or, which is also interesting, to hashing/caching (parts of) the universe.

Comment (fitzgen) [reply src/game-of-life/time-profiling.md]: This is an exercise in the `implementing.md` chapter! :)

> * Representing each cell with a byte makes iterating over cells easy, but it
>   comes at the cost of wasting memory. Each byte is eight bits, but we only
>   require a single bit to represent whether each cell is alive or dead. Refactor
>   the data representation so that each cell uses only a single bit of space.

File: src/game-of-life/implementing.md (hunk 30 above)
Comment (killercup) [reply src/game-of-life/implementing.md]: Ideally, we'd be able to have people new to x∈𝒫{Rust, WASM, HTML5 Canvas} read through this and tell us what (if anything) was confusing to them.

> I think we could skip the binary and start with sending a text representation across the boundary.

Good idea.

> This addresses the canvas unfamiliarity, but not the actual memory bits.

Doesn't it? If you as a first step expose a `Universe::render(&self) -> String`, you'll end up with a function that gives you, in JS, a string. No need to talk about memory there.

Then, you can have a section that replaces this impl with reading the raw memory (because you don't want to send stuff over from the Rust side), and after that you can introduce the canvas renderer on its own.

File: src/game-of-life/time-profiling.md (hunk 36 above)
Comment (killercup) [reply src/game-of-life/time-profiling.md]: Ohhh! I totally missed that! Should've read more carefully

File: src/game-of-life/code-size.md (hunk 35 above)
[the remark this answers is not in the document]
Comment (fitzgen) [reply src/game-of-life/code-size.md]: I seem to have the best results with `opt-level = "s"` in `Cargo.toml` and `wasm-opt -Oz`.

[the remark this answers is not in the document]
Comment (fitzgen) [reply src/game-of-life/code-size.md]: I added a comment mentioning that s sometimes beats z, and yet another reminder to always measure.

File: src/game-of-life/time-profiling.md (hunk 31 above)
Comment (fitzgen) [reply src/game-of-life/time-profiling.md]: Nice :)

File: src/game-of-life/code-size.md (hunk 32 above)
Comment (fitzgen) [reply src/game-of-life/code-size.md]: Done.

File: src/game-of-life/code-size.md (hunk 33 above)
Comment (fitzgen) [reply src/game-of-life/code-size.md]: Expanded this section on panicking a lot -- would appreciate if you took another look at it.

File: src/game-of-life/implementing.md (hunk 30 above)
Comment (fitzgen) [reply src/game-of-life/implementing.md]: This sounds good to me, but I also don't want to block landing on this change. I'll push it into a follow up. Thanks for digging into this :)

**Comment by fitzgen** [comment]
I'm going to go ahead and merge this PR now. This does *not* mean that the tutorial is complete, nor that it is perfect! I just think it is good enough to land now, and we can iterate in-tree from this point forward.

Thanks @alexcrichton and @killercup for giving this its first couple read throughs! :)
